frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/SelectionManager.java

```diff
@@ -46,185 +45,217 @@
-      } else {
-        LOGGER.logWarning("removeContainerRequest: Node is no longer a candidate: %s", nodeHost);
+  private void filterNodesByNodeLabel(String requestNodeLabel) {
+    if (requestNodeLabel != null) {
+      for (int i = filteredNodes.size(); i > 0; i--) {
+        Set<String> availableNodeLabels = allNodes.get(filteredNodes.get(i - 1)).getLabels();
+        if (!HadoopUtils.matchNodeLabel(requestNodeLabel, availableNodeLabels)) {
+          LOGGER.logDebug("NodeLabel does not match: Node: [%s] Request NodeLabel: [%s]",
+              filteredNodes.get(i - 1), requestNodeLabel);
+          filteredNodes.remove(i - 1);
+        }
       }
     }
   }
 
-  public synchronized SelectionResult select(
-      ResourceDescriptor requestResource, String requestNodeLabel, String requestNodeGpuType)
-      throws NotAvailableException {
-    LOGGER.logInfo(
-        "select: Given Request: Resource: [%s], NodeLabel: [%s], NodeGpuType: [%s]",
-        requestResource, requestNodeLabel, requestNodeGpuType);
-
-    // ClusterConfiguration is ready when this method is called, i.e. it is not null here.
-    ClusterConfiguration clusterConfiguration = am.getClusterConfiguration();
-    Map<String, NodeConfiguration> configuredNodes = clusterConfiguration.getNodes();
-
-    // Start to select from candidateNodes
-    SelectionResult selectionResult = null;
-    for (Node node : candidateNodes.values()) {
-      String nodeHost = node.getHost();
-      String logPrefix = String.format("select: [%s]: Test Node: ", nodeHost);
-      String rejectedLogPrefix = logPrefix + "Rejected: Reason: ";
-
-      LOGGER.logDebug(logPrefix + "Start: %s", node);
-
-      // Test NodeLabel
-      Set<String> availableNodeLabels = node.getLabels();
-      if (!HadoopUtils.matchNodeLabel(requestNodeLabel, availableNodeLabels)) {
-        LOGGER.logDebug(rejectedLogPrefix +
-                "NodeLabel does not match: Request NodeLabel: [%s], Available NodeLabel: [%s]",
-            requestNodeLabel, CommonExts.toString(availableNodeLabels));
-        continue;
-      }
-
-      // Test NodeGpuType
-      if (requestNodeGpuType != null) {
-        if (configuredNodes != null) {
+  private void filterNodesByGpuType(String requestNodeGpuType) {
+    if (requestNodeGpuType != null) {
+      Map<String, NodeConfiguration> configuredNodes = am.getClusterConfiguration().getNodes();
+      if (configuredNodes != null) {
+        for (int i = filteredNodes.size(); i > 0; i--) {
+          String nodeHost = filteredNodes.get(i - 1);
           if (!configuredNodes.containsKey(nodeHost)) {
-            LOGGER.logDebug(rejectedLogPrefix +
-                    "Node is not found in configured Nodes: Request NodeGpuType: [%s]",
-                requestNodeGpuType);
+            LOGGER.logDebug("Node:[%s] is not found in clusterConfiguration: Request NodeGpuType: [%s]", nodeHost, requestNodeGpuType);
+            filteredNodes.remove(i - 1);
```

Comment:
LOGGER.logDebug(rejectedLogPrefix +
                    "Node is not found in configured Nodes: Request NodeGpuType: [%s]",
                requestNodeGpuType);

Reply:
no big different.

---
In reply to: [174698625](https://github.com/Microsoft/pai/pull/318#discussion_r174698625) [](ancestors = 174698625)

Reply:
has big different. be careful pls 

---
In reply to: [175339816](https://github.com/Microsoft/pai/pull/318#discussion_r175339816) [](ancestors = 175339816,174698625)

Reply:
??

---
In reply to: [175410093](https://github.com/Microsoft/pai/pull/318#discussion_r175410093) [](ancestors = 175410093,175339816,174698625)